Replace sortable with draggable

Crabcyborg opened the pull request:
Fixes https://github.com/Strategy11/formidable-pro/issues/3175

The drag/drop code has been rewritten to work better with sections and field groups.

This also fixes a bug with deleted groups. When sections were deleted, the children fields weren't also being deleted when using the delete group option.

And it fixes a couple of items from https://github.com/Strategy11/formidable-pro/issues/3118 including:
- "I added a summary field and the grouping that happened is odd. It should wrap the page break and the summary separately."

It also appears to resolve this issue I can appear to replicate on master when I drop a field into a row really quickly. It won't sync my layout but will set the layout class value to a double class like `frm6 frm6`, looking like this as a result:
<img width="721" alt="Screen Shot 2022-09-15 at 4 26 49 PM" src="https://user-images.githubusercontent.com/9134515/190492106-7ae665f6-ddd0-4d04-b0d0-5fccdc37391b.png">

**Recording**
![QRx9245Dl8](https://user-images.githubusercontent.com/9134515/189980700-72bd9784-94ca-4870-9e61-1ac7724f6d66.gif)

**Completed**
- [x] Do we want class `this.classList.add( 'drop-me' );`?
- [x] A new dragged in field is not draggable.
- [x] Dragging does not work yet between rows.
- [x] Don’t lose any logic from updateFieldAfterMovingBetweenSections
    - [x] And insertNewFieldByDragging
- [x] Do I still need to maintain a $previousFieldContainer variable?
- [x] Only if the previous container had other sibling fields, remove the previous layout class.
- [x] Never place anything after the edit_field_type_end_divider.
- [x] Call allowDrop to make sure that some field types can’t be dragged into a section (see comment don't allow some field types inside section in sortable change function. It uses classes no-drop-placeholder and sortable-placeholder
- [x] Dragging a group into/out of a section should assign the whole group.
- [x] There are issues right now with dragging field groups.
- [x] There's code that tries to update a collapsed page when you drag a field. This doesn't appear to actually happen.
- [x] Page break+summary doesn’t init properly.
- [x] Reduce the negative margin on sections.
- [x] Trying to place a field beside itself back into it's original field group is weird when it in the only item in the group. The placeholder follows the edge of the field as it drags around.
- [x] Break into rows didn’t break into new groups properly
- [x] I can't move around a field within the group it was just in.
- [x] I don’t think sections properly delete in bulk-delete. The child fields seem to get left behind.
- [x] Right now it's calling frm_update_field_after_move too often (after after time a field is dragged). Only call it when section id changes.
- [x] It breaks when you load a long form with AJAX. The loaded items do not become draggable or droppable.
- [x] The field being dragged has an inconsistent z-index.

Crabcyborg commented:
@garretlaxton This isn't 100% finished yet but it's at a point where I could use your help testing it.

It's fairly likely there are still errors. I'd like your help tracking those down.

Some items that I know about are in the description.

Thanks!

garretlaxton commented:
I noticed a few things you have pointed out already, but here's some new things I see.
- [ ] Before you save the form, if you try to duplicate a group or section field it duplicates out of style.
https://www.loom.com/share/4130c9f69a214f2bb51e9fcf017fbdcc
- [ ] After you save the form, when you duplicate the group it puts it in the same group.
![localhost_form-ex_wp-admin_admin php_page=formidable frm_action=edit id=27](https://user-images.githubusercontent.com/16290508/189992186-a1e751b4-64c0-4cea-99fa-c1d50daeeed1.png)
- [ ] When groups are duplicated they duplicate in alphabetical order. Below, Section A & Section B should be next to each other as they were before duplicating.
![localhost_form-ex_wp-admin_admin php_page=formidable frm_action=edit id=27 (1)](https://user-images.githubusercontent.com/16290508/189993987-c255a22b-9dd9-4a05-8834-ff48bf66d7c3.png)
- [ ] After duplicating a section with a field, it still shows the `.frm_no_section_fields` border until you save or edit the section.
![localhost_form-ex_wp-admin_admin php_page=formidable frm_action=edit id=27 (2)](https://user-images.githubusercontent.com/16290508/189999458-e3d8a9fe-c90c-4789-8bb6-8e254f0ef327.png)

Crabcyborg commented:
Thanks @garretlaxton I've logged a couple of these issues so far as new issues. I think some of these could be handled later and were already issues prior to this update.

https://github.com/Strategy11/formidable-pro/issues/3816
https://github.com/Strategy11/formidable-pro/issues/3818

I haven't addressed the other items you mentioned yet either but I'll look into those soon.

I've pushed a few more updates though:

- There will be fewer network calls to update section IDs as it will only make a request when the sections actually change.
- I added a debounce so it syncs faster without hitting performance issues.
- When you delete a group with a section, the children fields of the section were not deleting, just the section itself.

garretlaxton commented:
How could I test the performance issue?

The first 2 look great! The fields sticking was a bit annoying,so glad that's solved!

Crabcyborg commented:
> How could I test the performance issue?

Sorry about the delay answering this @garretlaxton. I guess you could just test it with a lot of fields and sections, and if it's still fast, it's working. Testing that is less important than just making sure it all works and that it doesn't feel slow.

I've pushed a few more updates to this. I think this update should be good to go now.

Let me know if you can find any other issues! 🚀

garretlaxton commented:
I'm not sure if this is intended or related to this update, but sections are collapsed by default on the front-end when collapsible is selected.
![image](https://user-images.githubusercontent.com/16290508/193643224-69c14c1c-4624-4395-b36a-3cfc0b1eccd7.png)

I also notice section label positions are a bit messed up on the front-end as well.
- The right label still shows on the left and the section area is very small.
![image](https://user-images.githubusercontent.com/16290508/193645352-8083be75-a0f2-48a1-8b53-6317f4b47362.png)
- The none and hidden but leave space options hide the section completely.
![image](https://user-images.githubusercontent.com/16290508/193645259-d628ccd2-0822-4a69-b3f9-bbd6f0cb61f2.png)
- If you have collapsible selected and select the none or hidden but leave space option the whole section disappears including the field in the section

stephywells commented:
This is so much smoother now! Nice job @Crabcyborg !

I'm seeing a couple minor things.
1. When I hover a field over a section, it feels like it should go inside it on drop. Instead, it goes on the right where the drop line is showing.
<img width="859" alt="Screen Shot 2022-10-03 at 12 19 49 PM" src="https://user-images.githubusercontent.com/1116876/193650390-fefbde0a-5c1a-489c-a572-4e80cb26b38b.png">

2. The lines aren't quite lined up here:
<img width="822" alt="Screen Shot 2022-10-03 at 12 19 19 PM" src="https://user-images.githubusercontent.com/1116876/193650559-c08a9c48-64c5-4174-9971-ed60967eb80d.png">

@semelyak do you have any input on this? I think there might be related designs somewhere, but I'm not sure where to find them.

Crabcyborg commented on js/formidable_admin.js:
```diff
@@ -1388,23 +1466,24 @@ function frmAdminBuildJS() {
 		}
 
-		currentItem[0].addEventListener( 'click', function() {
-			maybeAddSaveAndDragIcons( this.dataset.fid );
```
@AbdiTolesa I removed this code. As far as I can tell, we didn't need this here. Just in the other place it was added. Can you confirm that?

Thanks!

AbdiTolesa replied:
@Crabcyborg I just pulled this update and tested it and found no issues so far after moving a field around sections and fields. Couldn't remember if that was needed really.

Crabcyborg commented:
>     1. When I hover a field over a section, it feels like it should go inside it on drop. Instead, it goes on the right where the drop line is showing.

Thanks @stephywells! I made the section title droppable now, and then swap out the droppable target so anything that goes over the section titles will try to enter the section's list. I also added a pointer-events: none style on the section header while the field is being dragged because I don't think the big blue outline helps either there.

> <img alt="Screen Shot 2022-10-03 at 12 19 49 PM" width="859" src="https://user-images.githubusercontent.com/1116876/193650390-fefbde0a-5c1a-489c-a572-4e80cb26b38b.png">

>     2. The lines aren't quite lined up here:

Thanks! I changed the padding in the section to fix that. It should look more centred now.

<img width="239" alt="Screen Shot 2022-10-04 at 11 52 09 AM" src="https://user-images.githubusercontent.com/9134515/193852345-754de0e1-3af8-489c-827a-31877ef3dbe4.png">

stephywells commented:
@Crabcyborg those are some weird issues. Let's log those and work on them in a separate PR. It seems like it would only make sense for sections to either show the title or not show it. The left, right, and inline stuff should probably all go.

Steve took a look at the changes here, and he would really like the field to be small again when dragging to make it easier to see where things will go. Can we add some css to the field being dragged so it looks more like adding a new field? We can use css to truncate the label, and set a width and bg color on it. It would be nice if we can show the field icon like we do when first dragging in a field. Is this easily to change?

Crabcyborg commented:
@stephywells What about when you're dragging a field group?

That's likely the weirdest looking use case (a large group), but in that case there is no "field type" icon unless we come up with one for groups (or use the row layout icon or something), or if we just grabbed one from one of the fields in the group.

stephywells commented:
I've got a couple small styling requests.

1. The drop zones are showing inside the box on the left, but outside the box on the right. Can the left one be outside the box too? 
<img width="855" alt="Screen Shot 2022-10-12 at 9 49 32 AM" src="https://user-images.githubusercontent.com/1116876/195391001-d4354b2e-1611-483b-974c-8a503d8574bf.png">

2. When moving a field, can we add a color overlay or something on it, so it's clear what is being moved? Like this:
<img width="900" alt="Screen Shot 2022-10-12 at 9 59 09 AM" src="https://user-images.githubusercontent.com/1116876/195414291-129995c3-0196-4f74-91f9-e30fe6dd95f1.png">

Crabcyborg commented:
> 1. The drop zones are showing inside the box on the left, but outside the box on the right. Can the left one be outside the box too?

@stephywells Good idea. I added a 4 pixel offset there. I'm already offsetting  in-between the fields in the group so it was pretty easy to change.

<img width="615" alt="Screen Shot 2022-10-12 at 4 08 35 PM" src="https://user-images.githubusercontent.com/9134515/195427493-8b4af546-f0ac-4991-ba92-5bf9d5d9b558.png">

>  2. When moving a field, can we add a color overlay or something on it, so it's clear what is being moved? Like this:

I'm not sure I totally follow the request but I made a few more adjustments.

I added a `box-shadow` to the draggable, and I set the item that you are trading to `opacity: 0.5` so it fades out, so it's more clear what you're actually dragging.

I also added another `pointer-events: none` to the field while dragging because I don't think it makes sense to have them show their hover effects while dragging - I noticed that from your screenshot and didn't like it. I've already done this for section headers but I guess I missed fields.

<img width="638" alt="Screen Shot 2022-10-12 at 4 09 56 PM" src="https://user-images.githubusercontent.com/9134515/195427760-de9f4a2f-9625-471c-b67a-fd415312d413.png">

stephywells commented:
1. Awesome!

So with #2, instead of showing the radio field, it would have a color block on top of it. Here it is in action. While dragging, the element being dragged changes in the background:
https://user-images.githubusercontent.com/1116876/195432955-4383157d-b952-4d92-b993-224a5d24c142.mp4

Crabcyborg commented:
@stephywells Thanks, I didn't know we were talking about how how Gutenberg handles it.

How does this look?

![ccjT4Fd51x](https://user-images.githubusercontent.com/9134515/195434948-e03dfc3e-5d1b-4a16-a401-f5ba7e363ef0.gif)

stephywells commented:
Yes! Exactly!! You came up with a way better solution for that than I did.

garretlaxton commented:
I'm noticing section fields can't be moved side by side after being placed in a form. I'm also having issues placing a field in a section when they are side by side. I am using the PR that was merged, not sure if I should be using this instead though.
https://www.loom.com/share/422cbb622ba542a8a0da2f01dc3276a1

Crabcyborg commented:
I think that's old code Garret. This look like caching to me.

The dragging item should look like the new field buttons now. It still looks like a section in your video which make some think it's not up to date.

garretlaxton commented:
Thanks for that. I was using the other PR that was merged so that was my issue. Switching to this worked great for me.
![Screenshot (21)](https://user-images.githubusercontent.com/16290508/195466514-e8cc992a-2e33-48ae-8fe5-10d2606b4b21.png)

The only issue I see is if you grab a new field, you can't put it back incase you grab the wrong field. You'll just have to delete it.
